Use shared directory iterator wrapper to allow build on Linux 6.x+

magnemahre opened the pull request:
The file_operations struct now uses the iterate_shared callback, instead of the old iterate callback. Linux provides a wrapper function to make an easy port.

Patch has only been tested with a couple of ODS-2 ISO images, loopback-mounted on a Linux 6.5 kernel.

magnemahre committed:
Use shared directory iterator wrapper to allow build on Linux 6.x+

magnemahre commented:
This patch addresses issue #2

magnemahre commented:
It doesn't quite cut it, it seems..  When a directory contains many files, it fails..

magne@fafnir:/mnt$ ls -la
ls: cannot access 'ARCHIE.DIR;1': No such file or directory
ls: cannot access 'ASKVAL28.DIR;1': No such file or directory
ls: cannot access 'BACKUP.SYS;1': No such file or directory
...
total 777
drwxr-xr-x  1 sync daemon  11264 Oct  1  1996  .
drwxr-xr-x 21 root root     4096 Dec  8 17:20  ..
-rwxr-xr-x  1 sync daemon   3538 Oct  1  1996 '$FREEWARE_README.TXT;8'
drwxr-xr-x  1 sync daemon  11264 Oct  1  1996 '000000.DIR;1'
drwxr-xr-x  1 sync daemon    512 Oct  1  1996 '000TOOLS.DIR;1'
-rwxr-xr-x  1 sync daemon  26026 Oct  1  1996 '00FREEWARE_ABSTRACT.TXT;1'
-rwxr-xr-x  1 sync daemon 729196 Oct  1  1996 'AAAREADME.V3;1'
drwxr-xr-x  1 sync daemon    512 Oct  1  1996 'ACC_CKPT010.DIR;1'
drwxr-xr-x  1 sync daemon    512 Oct  1  1996 'ACM47.DIR;1'
drwxr-xr-x  1 sync daemon    512 Oct  1  1996 'ADVENTURE.DIR;1'
drwxr-xr-x  1 sync daemon    512 Oct  1  1996 'ALTRIS.DIR;1'
drwxr-xr-x  1 sync daemon   1024 Oct  1  1996 'ANU-NEWS.DIR;1'
??????????  ? ?    ?           ?            ? 'ARCHIE.DIR;1'
??????????  ? ?    ?           ?            ? 'ASKVAL28.DIR;1'
??????????  ? ?    ?           ?            ? 'BACKUP.SYS;1'
...

I wonder if this was ever seen in the original code?

jayjwa commented:
It used to work before, up until the 'iterate' problem. I've used it on alot of images. Now I'm getting a different build error, with your changes:
```
/usr/src/linux-ods5/super.c: In function ‘ods5_read_inode’:
/usr/src/linux-ods5/super.c:319:58: error: ‘struct inode’ has no member named ‘i_ctime’; did you mean ‘i_atime’?
  319 |                 inode->i_mtime = inode->i_atime = inode->i_ctime = current_time(inode);
      |                                                          ^~~~~~~
      |                                                          i_atime
/usr/src/linux-ods5/super.c:324:24: error: ‘struct inode’ has no member named ‘i_ctime’; did you mean ‘i_atime’?
  324 |                 inode->i_ctime = v2utime(fi2->credate);
      |                        ^~~~~~~
      |                        i_atime
/usr/src/linux-ods5/super.c:329:24: error: ‘struct inode’ has no member named ‘i_ctime’; did you mean ‘i_atime’?
  329 |                 inode->i_ctime = v2utime(fi5->attdate);
      |                        ^~~~~~~
      |                        i_atime
make[2]: *** [scripts/Makefile.build:243: /usr/src/linux-ods5/super.o] Error 1
make[1]: *** [/usr/src/linux-6.6.6/Makefile:1913: /usr/src/linux-ods5] Error 2
make: *** [Makefile:234: __sub-make] Error 2
make: Leaving directory '/usr/src/linux-6.6.6'
```

magnemahre commented:
On 1/1/24 19:39, jayjwa wrote:
> error: ‘struct inode’ has no member named ‘i_ctime’;

This is a separate issue, due to a change in linux 6.6:

https://github.com/torvalds/linux/commit/13bc24457850583a2e7203ded05b7209ab4bc5ef

--Magne

magnemahre committed:
Use accessor functions for inode ctime manipulation

Linux kernel 6.6 renamed inode.i_ctime to __i_ctime, and recommended
the use of  accessor functions to access and manipulate it.

magnemahre commented:
On 1/1/24 19:39, jayjwa wrote:
> It used to work before, up until the 'iterate' problem. I've used it on 
> alot of images. Now I'm getting a different build error, with your changes:

I created a patch to use the new accessor functions. Note that I haven't 
had time to test it with 6.6-kernel myself yet, but I think it should 
work just fine

https://github.com/magnemahre/linux-ods5/commit/410af71eb9fed477e8d293a24980c71f173a2b6f

--Magne

jayjwa commented:
6.6.x update fixed the compile issue, but now I'm stuck where you are. 

```
# ls -l /mnt/hd
ls: cannot access '/mnt/hd/INDEXF.SYS;1': No such file or directory
...
total 52K
dr-xr-xr-x 1 lp bin 1.5K Nov  6  2007 '000000.DIR;1'/
dr-xr-xr-x 1 lp bin  512 Nov  6  2007 'ABSA043.DIR;1'/
-rwxr-x--- 1 lp bin    0 Nov  6  2007 'BACKUP.SYS;1'*
-rwxr-x--- 1 lp bin    0 Nov  6  2007 'BADBLK.SYS;1'*
-rwxr-x--- 1 lp bin    0 Nov  6  2007 'BADLOG.SYS;1'*
-rwxr-x--- 1 lp bin  21K Nov  6  2007 'BITMAP.SYS;1'*
dr-xr-xr-x 1 lp bin  512 Nov  6  2007 'CDROM.DIR;1'/
-rwxr-x--- 1 lp bin    0 Nov  6  2007 'CONTIN.SYS;1'*
-rwxr-x--- 1 lp bin    0 Nov  6  2007 'CORIMG.SYS;1'*
dr-xr-xr-x 1 lp bin  512 Nov  6  2007 'DECRAM02322.DIR;1'/
dr-xr-xr-x 1 lp bin  512 Nov  6  2007 'DXDV031.DIR;1'/
dr-xr-xr-x 1 lp bin  512 Nov  6  2007 'HSM043.DIR;1'/
?????????? ? ?  ?      ?            ? 'INDEXF.SYS;1'
?????????? ? ?  ?      ?            ? 'MR035.DIR;1'
?????????? ? ?  ?      ?            ? 'MRGATE035.DIR;1'
?????????? ? ?  ?      ?            ? 'MRIF035.DIR;1'
?????????? ? ?  ?      ?            ? 'MRX035.DIR;1'
?????????? ? ?  ?      ?            ? 'MTA030.DIR;1'
?????????? ? ?  ?      ?            ? 'MTAC020.DIR;1'
?????????? ? ?  ?      ?            ? 'SECURITY.SYS;1'
?????????? ? ?  ?      ?            ? 'SNA3270017.DIR;1'
?????????? ? ?  ?      ?            ? 'SNACSA022.DIR;1'
?????????? ? ?  ?      ?            ? 'SNACST013.DIR;1'
?????????? ? ?  ?      ?            ? 'SNADTFS034.DIR;1'
?????????? ? ?  ?      ?            ? 'SNADTFU034.DIR;1'
?????????? ? ?  ?      ?            ? 'SNALU0026.DIR;1'
?????????? ? ?  ?      ?            ? 'SNALU62025.DIR;1'
?????????? ? ?  ?      ?            ? 'SNAPRE015.DIR;1'
?????????? ? ?  ?      ?            ? 'SNAPU5021.DIR;1'
?????????? ? ?  ?      ?            ? 'SNARJE017.DIR;1'
?????????? ? ?  ?      ?            ? 'SNATE018.DIR;1'
?????????? ? ?  ?      ?            ? 'VOLSET.SYS;1'
?????????? ? ?  ?      ?            ? 'XMRC012.DIR;1'
```

hackerb9 commented:
With a caveat that I know *nothing* about VMS file systems, I think I may have found a clue.

Compiling with debugging enabled showed that the problem is related to the VBN. The first 512 bytes are read fine, but once `ods5_readdir()` detects that the current VBN has no more records, any further files found are screwy. The filenames are clearly being read correctly by od5_readdir, but they somehow aren't saved and `ods5_find_match()` is called with random binary strings from memory as the filenames. Since ods5_find_match is what looks up the attributes, we get the effect where `ls -l` shows question marks. 

By the way, I also tried compiling the module with a mutex in od5_read_dir instead of the macro for `.iterate_shared` and the problem still occurs, so the fact that the directory iterator changed in Linux 6 would seem to be a red herring.
